sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreIntegrationSuite.scala

```diff
@@ -222,6 +375,456 @@ class RocksDBStateStoreIntegrationSuite extends StreamTest
     }
   }
 
+  testWithChangelogCheckpointingEnabled(s"checkpointFormatVersion2") {
```

Comment:
From what I can tell, none of the new suites that were added cover the edge case in the design doc, right? There's no speculative execution here.

I _think_ what you could do is create new manual StateStores that simulate the race [here](https://docs.google.com/document/d/1uWRMbN927cRXhSm5oeV3pbwb6o73am4r1ckEJDhAHa0/edit), without needing to write a query that does this. Right?